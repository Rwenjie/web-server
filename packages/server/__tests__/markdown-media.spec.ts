import MarkdownIt from "markdown-it/lib";
import MediaPlugin from "../lib/markdown-media";
import Token from "markdown-it/lib/token";

describe("tokenizer", () => {
	let token: Token;

	const markdownIt = new MarkdownIt();
	markdownIt.use(MediaPlugin);

	markdownIt.renderer.rules.media = (t, i) => {
		token = t[i];
		return "No render result for tokenizer test";
	}

	it("should parse type, label, and href", () => {
		markdownIt.render("@gif[A gif video](/video/foo.mp4)");
		expect(token.tag).toBe("gif");
		expect(token.content).toBe("A gif video");
		expect(token.attrGet("src")).toBe("/video/foo.mp4");
	});

	it("should allow empty label and href", () => {
		markdownIt.render("@gif[]()");
		expect(token.content).toBe("");
		expect(token.attrGet("src")).toBe("");
	});

	it("should support escape ]", () => {
		markdownIt.render("@gif[A [gif\\] video]()");
		expect(token.content).toBe("A [gif] video");
	});

	it("should support escape )", () => {
		markdownIt.render("@gif[](/video/foo(bar\\).mp4)");
		expect(token.attrGet("src")).toBe("/video/foo(bar).mp4");
	});
});

describe("renderer", () => {
	const markdownIt = new MarkdownIt();
	markdownIt.use(MediaPlugin);

	it("should cooperate with others", () => {
		const markdown = `
text before

@gif[A gif video](/video/foo.mp4)

text after
`;
		expect(markdownIt.render(markdown)).toMatchSnapshot();
	});

	it("should render gif video", () => {
		expect(markdownIt.render("@gif[A gif video](/video/foo.mp4)")).toMatchSnapshot();
	});

	it("should render video", () => {
		expect(markdownIt.render("@video[/poster.png](/video/foo.mp4)")).toMatchSnapshot();
	});

	it("should escape html", () => {
		expect(markdownIt.render('@video[/f"o"o](/bar?a=b&c=d)')).toMatchSnapshot();
	});

	// 因为我用的 MarkdownIt 自带的 normalizeLink & validateLink，所以只测一种形式避免忘记检查
	it("should avoid XSS attack", () => {
		expect(markdownIt.render("@gif[](javascript:alert(1\\);)")).toMatchSnapshot();
		expect(markdownIt.render("@gif[<script>alert(1)</script>]()")).toMatchSnapshot();
		expect(markdownIt.render("@video[](javascript:alert(1\\);)")).toMatchSnapshot();
		expect(markdownIt.render("@video[javascript:alert(1\\);]()")).toMatchSnapshot();
	});
});
