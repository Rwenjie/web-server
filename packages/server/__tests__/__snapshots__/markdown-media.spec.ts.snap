// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should avoid XSS attack 1`] = `"<video title=\\"\\" src=\\"\\" loop muted></video>"`;

exports[`renderer should avoid XSS attack 2`] = `"<video title=\\"&lt;script&gt;alert(1)&lt;/script&gt;\\" src=\\"\\" loop muted></video>"`;

exports[`renderer should avoid XSS attack 3`] = `"<video src=\\"\\" controls></video>"`;

exports[`renderer should avoid XSS attack 4`] = `"<video src=\\"\\" controls></video>"`;

exports[`renderer should cooperate with others 1`] = `
"<p>text before</p>
<video src=\\"/video/foo.mp4\\" loop muted></video><p>text after</p>
"
`;

exports[`renderer should escape html 1`] = `"<video src=\\"/bar?a=b&amp;c=d\\" poster=\\"/f&quot;o&quot;o\\" controls></video>"`;

exports[`renderer should render gif video 1`] = `"<video title=\\"A gif video\\" src=\\"/video/foo.mp4\\" loop muted></video>"`;

exports[`renderer should render unknown type 1`] = `"[Unknown media type: unknown]"`;

exports[`renderer should render video 1`] = `"<video src=\\"/video/foo.mp4\\" poster=\\"/poster.png\\" controls></video>"`;
